# Classical Machine Learning Assignment

In [2]:
import pandas as pd
pd.set_option('display.precision', 2)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)


I import `pandas`, a library for the manipulation of tabulated data in Machine Learning. I set a few options including putting `max_columns` to `None` so that all of them can be seen in the Jupyter notebook without having to select subsets of columns. This will mean the Jupyter notebook will be easier to read but that the pdf will have some columns cropped out. Please see, the Jupyter notebook if you want to see the full data displayed.

In [3]:
df = pd.read_csv('CML_2024-09-09#3-wheat_genome_1.csv')
df

ZYTI                WIBD                KVJI                CSTT  \
0     7.12e+07   66780003.23249369    98615834.8241663   52159482.56117151   
1     5.80e+07   52206572.93108849   98309387.02801965  64901682.897473596   
2     8.93e+07  59058481.656934984   90340950.98559198   68251937.91355611   
3     4.16e+07   87279722.31068213  101620049.21358526   90873708.86792825   
4     9.06e+07  57388025.488219306   57568362.19853688   75526475.58607598   
...        ...                 ...                 ...                 ...   
4995  7.41e+07   58262507.06873764    92701179.7179384   67796505.77780084   
4996  6.63e+07  44913372.921941906   93559494.10271989    54065530.2835989   
4997  9.74e+07  51594995.715354934   65552552.49277553   98008359.76560771   
4998  5.86e+07   66843238.02653567    71275786.0738799   55771212.44004866   
4999  5.58e+07   76902968.80040008   82714128.91204472   55603146.54898293   

                   OFSQ      RXUY      GWEE      NANX      PMLC  \
0     82784798.00617953  6.06e+07  7.50e+07  8.37e+07  5.02e+07   
1     57695764.36746716  6.96e+07  9.30e+07  9.51e+07  6.38e+07   
2     89272597.48780008  6.70e+07  8.26e+07  9.48e+07  6.09e+07   
3     94655232.50324139  4.40e+07  7.62e+07  6.79e+07  4.83e+07   
4     60713729.34205614  8.32e+07  8.56e+07  8.17e+07  9.73e+07   
...                 ...       ...       ...       ...       ...   
4995  60996313.33816352  1.05e+08  3.67e+07  6.85e+07  9.24e+07   
4996  56083538.15939477  6.06e+07  7.73e+07  9.62e+07  6.80e+07   
4997  76960297.88906516  8.18e+07  6.96e+07  8.30e+07  8.47e+07   
4998  72847960.13200027  1.11e+08  5.37e+07  6.26e+07  9.16e+07   
4999  75568004.44787523  6.76e+07  7.99e+07  9.47e+07  4.97e+07   

                    JIEK      GXSO      PTAX                RLCO  \
0      67148304.13857159  7.71e+07  8.27e+07   68347063.23106696   
1      91303795.85450855  1.12e+08  9.82e+07  64604823.260036215   
2      53998583.56195885  6.46e+07  8.04e+07   59588103.00683621   
3      68304099.42017053  8.87e+07  1.00e+08   70052071.93627621   
4      78009778.11793171  5.50e+07  5.34e+07    50981332.5625033   
...                  ...       ...       ...                 ...   
4995   53654575.16920183  1.07e+08  9.00e+07   85839241.93596187   
4996   99407331.31693117  9.73e+07  9.74e+07   61454056.53538433   
4997   90739941.20183507  4.73e+07  7.65e+07   93275227.52404302   
4998   58169036.77614904  9.01e+07  8.24e+07   84858524.17996946   
4999  61735599.656930335  6.81e+07  9.00e+07   70151739.14748453   

                   RNOZ      QRWO               PSSJ                WBYY  \
0     86863631.66238013  5.14e+07  76935314.59720963  105443191.56660278   
1     81374620.04454753  5.88e+07  89711575.45912442   39303004.81084073   
2     86360346.40430261  7.96e+07  86348427.75161901    99173173.9837645   
3     68627709.75841182  7.84e+07  57881674.17562354   71188279.49110305   
4     65076827.86445975  5.69e+07  97149817.50893311  58031957.693665884   
...                 ...       ...                ...                 ...   
4995  77902522.01750344  5.84e+07  86443114.50503628   57347972.89513925   
4996  89855424.75312191  5.03e+07  71468658.62508146   40146847.83650946   
4997   52439119.3751166  5.20e+07  72569921.86725812    71439253.9880689   
4998  78123267.90743856  6.47e+07  95093568.50489672   64316317.91820028   
4999  82451061.22957435  5.02e+07   79849349.2104344   95368848.53425784   

          BDGK      FWKW      PFKF      ZKOY      XBUG      OMXX      IXKB  \
0     8.91e+07  7.79e+07  6.44e+07  8.09e+07  6.33e+07  9.81e+07  8.40e+07   
1     5.97e+07  5.60e+07 -1.33e+11  8.82e+07  6.01e+07  6.57e+07  1.06e+08   
2     9.37e+07  7.78e+07  5.12e+07  8.20e+07  5.92e+07  9.50e+07  7.23e+07   
3     1.01e+08  5.18e+07  6.28e+07  6.96e+07  6.93e+07  5.59e+07  7.63e+07   
4     9.69e+07  1.02e+08  7.51e+07  9.08e+07  8.59e+07  8.57e+07  7.25e+07   
...        ...       ...       ...       ...       ...       ...      

I read the Landrace Wheat dataset into pandas and through a sampling of the first 5 and last 5 rows I can see there are 40 genetic marker features and 5 response measurements. It looks like the features are floating point numbers in the millions, and the responses floating point numbers in the billions however given there are 5000 rows, too many to manually inspect I need other ways to better understand the data.

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 45 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   ZYTI     5000 non-null   float64
 1   WIBD     4999 non-null   object 
 2   KVJI     5000 non-null   object 
 3   CSTT     5000 non-null   object 
 4   OFSQ     4998 non-null   object 
 5   RXUY     4999 non-null   float64
 6   GWEE     5000 non-null   float64
 7   NANX     4999 non-null   float64
 8   PMLC     4998 non-null   float64
 9   JIEK     5000 non-null   object 
 10  GXSO     4999 non-null   float64
 11  PTAX     5000 non-null   float64
 12  RLCO     5000 non-null   object 
 13  RNOZ     5000 non-null   object 
 14  QRWO     5000 non-null   float64
 15  PSSJ     5000 non-null   object 
 16  WBYY     4999 non-null   object 
 17  BDGK     5000 non-null   float64
 18  FWKW     4999 non-null   float64
 19  PFKF     5000 non-null   float64
 20  ZKOY     4998 non-null   float64
 21  XBUG     5000 

Using `.info` I can get a summary of the dataframe where I see that my initial impression of the data was inaccurate. Viewing as a scrollable element allows me to see all the data.

Firstly, I can see that some feature columns have `null` data such as `MHEC` which has 3 `null` data and 4997 `non-null`, not evident from our initial 10 row sampling of the data. Similarly, for the reponse variables `RESP__0` has 1 `null` and 4999 `non-null`. 

Secondly, in addition to the 29 `float64` numbers I also have 16 `object` data types in both the marker feature such as `TUXT` and responses `RESP__1`. Where `object` is a catch-all data type (`dtype`) which occurs when 'multiple dtypes in a single column, the dtype of the column will be chosen to accommodate all of the data types (object is the most general).'[1]. 

[1] https://pandas.pydata.org/pandas-docs/stable/user_guide/basics.html#basics-dtypes

In [5]:
for col in df.columns:
    if df[col].dtype == 'object':
        print(df[col].apply(type).value_counts())


WIBD
<class 'str'>      4999
<class 'float'>       1
Name: count, dtype: int64
KVJI
<class 'str'>    5000
Name: count, dtype: int64
CSTT
<class 'str'>    5000
Name: count, dtype: int64
OFSQ
<class 'str'>      4998
<class 'float'>       2
Name: count, dtype: int64
JIEK
<class 'str'>    5000
Name: count, dtype: int64
RLCO
<class 'str'>    5000
Name: count, dtype: int64
RNOZ
<class 'str'>    5000
Name: count, dtype: int64
PSSJ
<class 'str'>    5000
Name: count, dtype: int64
WBYY
<class 'str'>      4999
<class 'float'>       1
Name: count, dtype: int64
TUXT
<class 'str'>      4999
<class 'float'>       1
Name: count, dtype: int64
ROQE
<class 'str'>      4998
<class 'float'>       2
Name: count, dtype: int64
ZMHO
<class 'str'>    5000
Name: count, dtype: int64
RTDN
<class 'str'>    5000
Name: count, dtype: int64
MJXP
<class 'str'>    5000
Name: count, dtype: int64
RESP__0
<class 'str'>      4999
<class 'float'>       1
Name: count, dtype: int64
RESP__1
<class 'str'>    5000
Name: count, dty

I investigate all the columns that have `dtype` `object` and gettings counts of all the types in that column. Note that although the column has to be a type that encompasses all other types, the individual values still preserve more nuanced types

First, I have those which are mixed types such as `str` and `float`. For example, `WIBD` has 4999 `str` and one `float`. The obvious hypothesis is that the `null` values are being interpreted as floats as we can see in `df.info` it has a count of 4999 non-null values.

In [6]:
for col in df.columns:
    if df[col].dtype == 'object':
        val_types = df[col].apply(type)
        if val_types.nunique() > 1:
            float_indices = df.index[df[col].apply(type) == float].tolist()
            null_indices = df.index[df[col].isnull()].tolist()
            floats_same_nulls = float_indices == null_indices
            print(f"{col}: {floats_same_nulls}, {float_indices} floats, {null_indices} nulls")

WIBD: True, [1032] floats, [1032] nulls
OFSQ: True, [191, 2187] floats, [191, 2187] nulls
WBYY: True, [215] floats, [215] nulls
TUXT: True, [3138] floats, [3138] nulls
ROQE: True, [1738, 2247] floats, [1738, 2247] nulls
RESP__0: True, [1644] floats, [1644] nulls


I confirm this hypothesis by showing that all the mixed columns have the float types in exactly the same places as where the nulls we saw in `df.info` were found i.e. it is only an `object` because we have non-handled the `null` yet. For example, `WIBD` has a single float value at index `1032` which is also its `null` index.

Second, there is the question of what to do with the strings. The reason that pandas sometimes casts to strings (technically `object`) is that it cannot convert to floats as they do not provide enough precision[2]. 

However, I cannot use strings to analyse the data so I have a choice between casting to float anyway or perhaps using `Decimal`. The advantage of the former is that `pandas` uses `numpy` under the hood which is optimised for floats including lower-memory footprint, hardward supported with vectorised operations and it allows us to use functions which work with floats such as `df.describe`. The advantages of the latter is exact precision which in some contexts might matter.

[2] https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html#type-inference

In [9]:
def safe_float(x):
    try:
        return float(x)
    except ValueError:
        return x  

for col in df.columns:
    if df[col].dtype == 'object':
        col_float = df[col].apply(lambda x: safe_float(x) if not pd.isna(x) else x)
        non_float_mask = ~col_float.apply(lambda x: isinstance(x, float) or pd.isna(x))
        
        if non_float_mask.any():
            print(f"\nColumn: {col}")
            non_float_vals = df[col][non_float_mask].index
            print(col_float[non_float_mask].tolist())


Column: WIBD
['<Err>']

Column: KVJI
['<Err>']

Column: CSTT
['<Err>', '<Err>']

Column: OFSQ
['<Err>']

Column: JIEK
['<Err>']

Column: RLCO
['<Err>']

Column: RNOZ
['<Err>']

Column: PSSJ
['<Err>']

Column: WBYY
['<Err>']

Column: TUXT
['<Err>']

Column: ROQE
['<Err>']

Column: ZMHO
['<Err>']

Column: RTDN
['<Err>', '<Err>']

Column: MJXP
['<Err>', '<Err>']

Column: RESP__0
['<Err>']

Column: RESP__1
['<Err>', '<Err>']


In order to use the data I try casting to float. Initially, I tried the simplest approach which is to do `df['col'].astype(float)`. To understand why I use a `safe_float` function which simply wraps each `float` cast in a `try-except` block. I then use this method as a `lambda` to each value in the column noting that those that fail will be left as `x`, which is important as I want to be able to see the data that couldn't be casted to understand the dataset better.

Then, I create a mask to find the indices of values in the columns that are neither `null` which we have already investigated or successfully cast to `float`, i.e. to find the failed casts. Printing them out for each column, I fine that it is the same problem of `<Err>`

In [8]:
from decimal import Decimal

def safe_float(x):
    try:
        return float(x)
    except (ValueError, TypeError):
        return x
    
def safe_decimal(x):
    try:
        return Decimal(x)
    except (ValueError, TypeError):
        return x

for col in df.columns:
    if df[col].dtype == 'object':
        col_float = df[col].apply(safe_float)
        col_decimal = df[col].apply(safe_decimal)

        non_float = col_float[~col_float.apply(lambda x: isinstance(x, float))]
        if not non_float.empty:
            print(f"{col}: {non_float}")
        non_decimal = col_decimal[~col_decimal.apply(lambda x: isinstance(x, Decimal))]
        if not non_decimal.empty:
            print(f"{col}: {non_decimal}")
        
        float_decimal_diff = col_float != col_decimal
        if float_decimal_diff.any():
            diff_indices = float_decimal_diff[float_decimal_diff].index.tolist()
            float_diffs = col_float.loc[diff_indices]
            decimal_diffs = col_decimal.loc[diff_indices]
            print(f"Float diffs in {col}: {float_diffs.tolist()} == {decimal_diffs.tolist()}")


InvalidOperation: [<class 'decimal.ConversionSyntax'>]

In [ ]:
df.describe()

ZYTI      RXUY      GWEE      NANX      PMLC      GXSO      PTAX  \
count  5.00e+03  5.00e+03  5.00e+03  5.00e+03  5.00e+03  5.00e+03  5.00e+03   
mean   1.09e+08  1.14e+08  4.80e+07  7.52e+07  9.58e+07  5.23e+07  3.93e+07   
std    3.70e+09  4.24e+09  1.90e+09  3.62e+09  1.55e+09  3.50e+09  3.86e+09   
min   -1.34e+11 -1.54e+11 -1.35e+11 -1.86e+11  3.15e+07 -1.50e+11 -1.82e+11   
25%    6.25e+07  5.82e+07  5.86e+07  6.45e+07  5.69e+07  5.67e+07  6.28e+07   
50%    7.17e+07  6.75e+07  7.67e+07  7.71e+07  7.31e+07  7.04e+07  8.11e+07   
75%    8.29e+07  8.65e+07  8.97e+07  8.91e+07  9.09e+07  9.38e+07  9.29e+07   
max    1.79e+11  1.89e+11  1.20e+08  1.76e+11  1.10e+11  1.58e+11  1.36e+11   

           QRWO      BDGK      FWKW      PFKF      ZKOY      XBUG      OMXX  \
count  5.00e+03  5.00e+03  5.00e+03  5.00e+03  5.00e+03  5.00e+03  5.00e+03   
mean   6.10e+07  4.39e+07  1.42e+08  4.43e+07  1.11e+08  9.22e+07  2.02e+08   
std    1.18e+07  2.68e+09  3.56e+09  1.88e+09  2.13e+09  4.81e+09  4.52e+09   
min    0.00e+00 -1.89e+11  3.20e+07 -1.33e+11  5.13e+07 -1.58e+11  3.91e+07   
25%    5.25e+07  6.37e+07  5.90e+07  5.59e+07  7.28e+07  5.85e+07  6.39e+07   
50%    6.02e+07  8.76e+07  6.74e+07  6.72e+07  8.00e+07  6.63e+07  7.68e+07   
75%    6.94e+07  9.70e+07  8.05e+07  8.74e+07  8.94e+07  7.57e+07  8.64e+07   
max    1.06e+08  1.25e+08  1.79e+11  1.25e+08  1.50e+11  1.76e+11  1.91e+11   

           IXKB      YKJO      GQUZ      VKRW      DDTT      QXYM      MHEC  \
count  5.00e+03  5.00e+03  5.00e+03  5.00e+03  5.00e+03  5.00e+03  5.00e+03   
mean   7.04e+07  7.64e+07  1.06e+08  5.05e+07  7.97e+07  5.30e+07  7.36e+07   
std    3.26e+09  2.66e+09  2.55e+09  1.61e+09  2.80e+09  3.91e+09  3.81e+09   
min   -1.84e+11 -1.17e+11  0.00e+00 -1.14e+11 -1.35e+11 -1.60e+11 -1.89e+11   
25%    6.88e+07  6.07e+07  5.91e+07  6.03e+07  6.21e+07  6.91e+07  5.49e+07   
50%    8.06e+07  7.06e+07  7.02e+07  7.19e+07  7.85e+07  7.88e+07  7.27e+07   
75%    9.01e+07  7.91e+07  8.13e+07  8.69e+07  9.25e+07  8.60e+07  9.12e+07   
max    1.39e+11  1.47e+11  1.80e+11  1.19e+08  1.45e+11  1.79e+11  1.92e+11   

           UZSD      AOQU      RJWA      VTQI      XYKI   RESP__2   RESP__3  \
count  5.00e+03  5.00e+03  5.00e+03  5.00e+03  5.00e+03  5.00e+03  5.00e+03   
mean   6.55e+07  4.24e+07  8.95e+07  8.58e+06  2.19e+06  1.70e+10  3.25e+10   
std    1.64e+07  2.78e+09  1.69e+09  4.97e+09  4.91e+09  4.46e+09  5.45e+09   
min    0.00e+00 -1.26e+11  3.00e+07 -1.98e+11 -1.96e+11 -1.72e+11  2.18e+10   
25%    5.34e+07  5.72e+07  5.49e+07  5.89e+07  6.82e+07  1.59e+10  2.85e+10   
50%    6.16e+07  6.71e+07  6.31e+07  6.74e+07  7.67e+07  1.72e+10  3.09e+10   
75%    7.65e+07  8.24e+07  7.35e+07  7.72e+07  8.42e+07  1.84e+10  3.65e+10   
max    1.20e+08  1.02e+11  1.20e+11  1.28e+11  1.93e+11  1.27e+11  1.08e+11   

        RESP__4  
count  5.00e+03  
mean   3.49e+10  
std    6.03e+09  
min    2.03e+10  
25%    3.00e+10  
50%    3.50e+10  
75%    3.86e+10  
max    1.51e+11

Another way we can explore the data is using `.describe` which calculates common statistical metrics like the `mean`.

Our initial visual inspection of the data gave the impression that features are largely in the millions 

## 1. Is the wheat in this dataset homogeneous, random or does it fall into distinct types?

2. If the wheat is homogenenous or random, then identify the overall characteristics of the complete sample.

3. If there are distinct types of wheat in the dataset, then identify the characteristics of each type of wheat.

4. What is the relationship between the input variables (first 40 features) and the 5 response variables?

5. Are there specific features which ‘drive’ specific responses?
    a) It so, how?
    b) What are the important features and what is their relative impact?

6. If there are different types of wheat present I the dataset, then which type of wheat has the greatest impact on each response variable?